Feature/uncertainty

chahank opened the pull request:
NOTE: THIS IS NOW A FULL REVIEW REQUEST.

--- Pre-review ---
This is the second version of the uncertainty module. For this review, I would appreciate receiving general feedback on the architecture choices made in this module. To be clear: I DO NOT EXPECT A FULL REVIEW. 
For instance, the module is already 984 lines long (with all coments etc). But it is expected that more features will be added. Is this ok or not? @emanuel-schmid is it ok to add the dependency `SALib` to CLIMADA?

While some rudimentary testing was done to see whether the code actually runs through, no rigorous tests of whether the correct numerical data is generated has been done so far. This will be done soon, but since architecture changes might still come from your review, not right now. Furthermore, the example (tutorial) is very rudimentary, just so to illustrate the work flow of the module. 

Compared to the first version of this module, there are several changes. The module now consists of two main classes and three child classes.

- `UncVar`: to describe uncertainty variables
- `Uncertainty`: main uncertainty class with generic sampling and sensitivity analysis methods
    - `UncImpact`: uncertainty with the specific metrics from Impact.calc
    - `UncCostBenefit`: uncertainty with the specific metrics from CostBenefit.calc
    - `UncRobustness`: not yet implemented (for after the march release)

The reason for separating `Uncertainty` into two child classes is that the conversion of the ouput from `Impact.calc` and `CostBenefit.calc` to `pandas.Dataframe` that are used to conduct the statistical analysis of the sampled data (mean, variance, sensitivity, ...) is very specific. In principle, one could also simply use specific methods inside of the mother class. However, I felt like this is a cleaner implementation. It has also the added value that classes do not become too large, and it is possible to add for instance particular plotting methods. The class `UncRobustness` also then neatly fits in as a class in which an automatic noise test is done on the input data.

Thank you in advance for your time and your inputs!

aleeciu commented:
Hi Chahan,
great work!! 

I have gone through the module and I think it is very readable, excpet for lines 313-318 (and similar calls below too). Are you there importing submodules from Salib? Why not just importing them on top when importing all modules?

In terms of architecture, I agree with the strucuture. I would perhaps propose to move UncImpact, UncCostBenefit to a different file. 

Also, I haven't use the CostBenefit analysis much so I might be wrong, but why would we need an ad hoc uncertainty class for it? at the end of the day, to do a CBA you need to run impact in various ways, so would it be possible to only have the UncImpact class which optionally gets the entity_future argument, in which case outputs from the CBA are returned? Also, maybe it also makes sense to add a cost uncertainty argument..

Thanks again..

chahank commented:
Dear Alessio,
Thanks for the feedback! It is very similar to what Emanuel told me :D. 

> Are you there importing submodules from Salib? Why not just importing them on top when importing all modules?
These lines are for selecting the appropriate Salib function. In this way, if Salib adds another sampling function for instance, the code must most probably not be changed to accommodate for that. I thought this was cleaner than to separately import the salib methods and then use a dictionary to call them...

> In terms of architecture, I agree with the strucuture. I would perhaps propose to move UncImpact, UncCostBenefit to a different file.
Good idea, I will do that.

> Also, I haven't use the CostBenefit analysis much so I might be wrong, but why would we need an ad hoc uncertainty class for it? at the end of the day, to do a CBA you need to run impact in various ways, 
That is what I thought first too, but the `climada.engine.cost_benefit.calc` actually works quite differently from `climada.engine.impact.calc`. First, it works with entities instead of exposures/impact functions. Second, the outputs are quite different. Finally, the plots for cost benefit and impact are quite different, hence I thought having them in separate classes makes sense. But, Emanuel made the exact same remark. Thus, I will consider it further.

> Also, maybe it also makes sense to add a cost uncertainty argument..
Definitely! But I think they are already included as I return all the arguments from the `cost_benefit.calc` except the `freq_curve`. I will double check.

chahank commented:
This module is now ready for a first detailed review. It is possible to do an uncertainty analysis for both `impact` and `cost_benefit` computations. It is also possible to perform a subsequent sensitivity analysis using the module [salib](salib.readthedocs.io) for both `impact` and `cost_benefit`. The module also provides some basic plotting functions. A basic tutorial was written and extensive tests are running. 

The goal would be to include this module in the up-coming release (end of March / beginning of April). A thorough review is thus very highly appreciated. 

Note that the plotting function as well as the tutorial might still see some improvements in the coming weeks.,

chahank commented on climada/engine/uncertainty/unc_impact.py:
```diff
@@ -255,3 +256,29 @@ def plot_rp_distribution(self):
         return fig, ax
+    
+    
+    def plot_sensitivity_map(self, exp):
```
The latest commit add the possibility to plot a sensitivity map which shows at each exposure location what is the largest sensitivity index. 

Please note that this commit depends on the branch `feature/categorical_plot` which will be put to review soon.

emanuel-schmid commented:
> 
>  @emanuel-schmid is it ok to add the dependency `SALib` to CLIMADA?
> 
Should be ok. 👍 \
The dependencies are pandas, scipy, numpy and matplotlib, starting from versions way lower than what we use, and the overall size of all package files is < 10M.

Evelyn-M commented on climada/engine/uncertainty/unc_cost_benefit.py:
```diff
@@ -0,0 +1,227 @@
+# Future planed features:
+# - Add 'efc' (frequency curve) to UncCostBenenfit
+
+class UncCostBenefit(Uncertainty):
+    """
+    Cost Benefit Uncertainty analysis class
+
+    This is the base class to perform uncertainty analysis on the outputs of a
+    climada.engine.costbenefit.CostBenefit() object using the SALib package.
+
+    """
```
Can you provide a pseudo-code usage example? I find this really hard to grasp in terms of all the pre-steps that need to be done, and especially highlighting the fact that some measures need to be defined, computed and added as an unc_var to the entity, which in turn is an UncVar class. The measures are no-where explicitly mentioned in the docstrings (yes, even if it's a no-brainer that you need some for doing cost-benefit calcs).

chahank replied:
I would argue that is what the tutorial is for right?

Evelyn-M commented on climada/engine/uncertainty/unc_cost_benefit.py:
```diff
@@ -0,0 +1,227 @@
+            'ent_fut': UncVar.var_to_uncvar(ent_fut_unc)
+            }
+
+        metrics =  {
+            'tot_climate_risk': pd.DataFrame([]),
+            'benefit': pd.DataFrame([]),
+            'cost_ben_ratio': pd.DataFrame([]),
+            'imp_meas_present': pd.DataFrame([]),
+            'imp_meas_future': pd.DataFrame([])
+            }
```
no cost uncertainty by default?

chahank replied:
This is in the `imp_meas_present` and `imp_meas_future`, c.f. https://github.com/CLIMADA-project/climada_python/blob/26da4ccb775f38e52df2e9022ebf6e96e61ded64/climada/engine/cost_benefit.py#L104

Evelyn-M commented on climada/engine/uncertainty/unc_impact.py:
```diff
@@ -0,0 +1,309 @@
+                                           chunsize = chunksize)
+
+        else:
+            imp_metrics = map(self._map_impact_calc, self.sample.iterrows())
```
I've never seen the map - iterrows() combo. Why did you prefer this over the apply function?
https://stackoverflow.com/questions/19798153/difference-between-map-applymap-and-apply-methods-in-pandas

chahank replied:
In order to use `pathos`. But maybe I should rather use `dask`. I can see if it works too.

Evelyn-M commented on climada/engine/uncertainty/unc_impact.py:
```diff
@@ -0,0 +1,309 @@
+
+        """
+
+        # [1] only the rows of the dataframe passed by pd.DataFrame.iterrows()
```
the [1] hence shouldnt be necessary when using the apply instead of the iterrows() / map() syntax

chahank replied:
I do not see how you could use the apply method. Could you explain how you would do it? Then one could use `dask` for parallelization.

Evelyn-M replied:
`impact_metrics = self.samples_df.apply(_map_impact_calc)` ?

Evelyn-M commented on climada/engine/uncertainty/unc_impact.py:
```diff
@@ -0,0 +1,309 @@
+            imp_metrics = map(self._map_impact_calc, self.sample.iterrows())
+        
+        #Perform the actual computation
+        with log_level(level='ERROR', name_prefix='climada'):
+            [aai_agg_list, freq_curve_list,
+             eai_exp_list, at_event_list,
+             tot_value_list] = list(zip(*imp_metrics))
```
This doesnt look like it performs the actual calculation..

chahank replied:
That is why I added the comment. The `map` returns an iterator. The evaluation of this iterator happens in this line where it is cast into a list.

Evelyn-M replied:
ah. this is very tricky to spot. can you expand the comment in a way you just explained it here?

Evelyn-M commented on climada/engine/uncertainty/unc_impact.py:
```diff
@@ -0,0 +1,309 @@
+
+    def plot_rp_distribution(self):
+        """
+        Plot the distribution of return period values.
```
Perhaps one more sentence at what stage this plotting method works (i.e. what needs to be computed beforehand)

chahank replied:
I am not sure whether this makes too much sense now. For example, in the `Impact` class, there is no mention that you need to run `.calc` before you can plot... Why should one add such points in particular here? There is already a check with an error if the distribution data is not present

```
        if not self.metrics:
            raise ValueError("No uncertainty data present for these metrics. "+
                    "Please run an uncertainty analysis first.")
```

Evelyn-M replied:
yes, fair

Evelyn-M commented on climada/engine/uncertainty/unc_cost_benefit.py:
```diff
@@ -0,0 +1,227 @@
+
+        """
+
+        # [1] only the rows of the dataframe passed by pd.DataFrame.iterrows()
+        haz_samples = param_sample[1][self.unc_vars['haz'].labels].to_dict()
```
see analogous comment in unc_impact.py to this

Evelyn-M commented on climada/engine/uncertainty/unc_cost_benefit.py:
```diff
@@ -0,0 +1,227 @@
+            cb_metrics = map(partial(self._map_costben_calc, **kwargs),
+                             self.sample.iterrows())
+        
+        #Perform the actual computation
+        with log_level(level='ERROR', name_prefix='climada'):
+            [imp_meas_present,
+             imp_meas_future,
+             tot_climate_risk,
+             benefit,
+             cost_ben_ratio] = list(zip(*cb_metrics)) #Transpose list of list
```
see analogous comment in unc_impact to this

Evelyn-M commented on climada/util/value_representation.py:
```diff
@@ -14,16 +14,19 @@
 def sig_dig(x, n_sig_dig = 16):
     """
     Rounds x to n_sig_dig number of significant digits.
+    0, inf, Nan are returned unchanged.
     Examples: 1.234567 -> 1.2346, 123456.89 -> 123460.0
```
Is this true? Why would it acutally correctly round up 56 to 60 in the 2nd example, but wrongly round down 6.7e-5 to 6e-5 in the first example?

chahank replied:
I do not understand... This rounds your number to a certain number of significant digits. In the example, `n_sig_dig=5` (I added this for clarity). Basically,  it counts from left up to 5 digits, then rounds that last digit and sets the remaining values to the right to 0.

Evelyn-M reviewed (commented):
Very nice module that can do a lot of stuff!! Now only a comprehensive jupyter notebook, which documents and illustrates well a wider range of cases that this extensive module covers, would be missing.

chahank commented on doc/tutorial/climada_engine_uncertainty.ipynb:
```diff
@@ -0,0 +1,1821 @@
+    "import scipy as sp\n",
+    "import pandas as pd\n",
+    "import numpy as np\n",
+    "import seaborn as sns\n",
```
Good point.  But it makes nice plots! Maybe we should ask @emanuel-schmid : could we also add `seaborn` to the requirements?

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,685 @@
+        
+        if sampling_kwargs is None: sampling_kwargs = {}
+
+        #To import a submodule from a module use 'from_list' necessary
+        #c.f. https://stackoverflow.com/questions/2724260/why-does-pythons-import-require-fromlist
+        salib_sampling_method = getattr(
+            __import__('SALib.sample',
+                       fromlist=[sampling_method]
+                       ),
+            sampling_method
+            )
```
I understand that the code is rather complex, although I think the name is pretty explicit and can be read even without understanding what is going on exactly. Your proposal would imho be very bad coding style. Even if it already exists in the CLIMADA codebase, it should not. 

Your proposal is to hard-code the dictionary of methods in the salib sample module. But this is exactly what the above lines of code are doing (extract the dictionnary of salib methods). This is much much cleaner, as 1) we do not introduce very hard to maintain redundant code 2) it can keep up with potential new releases (for example, if `salib` adds a new sampling function, this code would automatically work with it).

In my view, one should always avoid hard-coded variables if not necessary as these are basically impossible to maintain. 

The above code is a bit more advanced python, thus the link to the stackoverflow thread. If needed, one could add extra comments.

Evelyn-M replied:
hmm yes. i guess since one does not necessarily need to look at this beast when setting the kwarg sampling method in the function call, the advantages of having it like this may outweigh the hard-coding but easy to read option.

chahank replied:
I tried to slightly improve the comment in f296d47664e1a6d792462d0ca9f4587e0fcca62e

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+    """
+    Uncertainty variable
+
+    An uncertainty variable requires a single or multi-parameter function.
+    The parameters must follow a given distribution.
```
How about I rename the `unc_var` attribute as `unc_fct` or `var_fct`  or `unc_func` or `uncf` ?

chahank replied:
Another possibility would be to rename the class. But I kind of like that UncVar is the name of the class that contains all the information about the uncertainty variable.

Evelyn-M replied:
yes, I think `unc_func` or `uncvar_func` is a pretty descriptive name and nice to distinguish from the class container

chahank replied:
I set the name to `uncvar_func` in e09a14270e5c17c33a9226f6b939466fa1f2b515

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+LOGGER = logging.getLogger(__name__)
+
+
+class UncVar():
```
Good points. Let's keep the discussino in the comment 
>Also here, this description is a bit short and rather focuses on unc_var than mentioning that UncVar is a container for unc_var and distr_dict. Those sentence could still remain as further explanation, then.
above.

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+            impf_set = ImpactFuncSet()
+            impf_set.append(imp_fun)
+            return impf_set
+        distr_dict = {"G": sp.stats.uniform(0.8, 1),
+              "v_half": sp.stats.uniform(50, 100),
+              "vmin": sp.stats.norm(loc=15, scale=30),
+              "k": sp.stats.randint(low=1, high=9)
+              }
```
Yes. I do not see any reasonable example where this would not be the case? What do you have in mind?

Evelyn-M replied:
What if you're pretty sure about one component for a specific case? Say I run a set of very local uncertainty analyses, and for one study area I just _know_ that vmin = 15m/s is a good starting point; then i may still want to define vmin as a parameter, for I might not be so certain about this in the other study areas where I also want to run uncertainty analyses.

chahank replied:
Well, say you have the following for the exposures
```
unc_exp_func(x_exp, exp=exp):
    exp.gdf.values *= x_exp
    return exp
```
Then, if you want to compute the impact with a constant `x_exp=1`, you just do
```
exp = unc_exp_func(1)
unc_imp = UncImpact(exp, impf_unc, haz_unc)
```
The `UncImpact` and `UncCostBenefit` accept either `UncVars` or directly the CLIMADA object as entry.

chahank replied:
Or, in the case of multiple variables define the function with a default value
```
unc_impf_func(G, k, v_half, v_min=1):
    ...
```

Evelyn-M replied:
ok, so if I set a default value, no distribution is expected later on, correct?

chahank replied:
Well, in fact all the parameters supplied in the dictionary will be replaced. So, say your function is
```
unc_func(x, y=10, z=1)
```
then `x` must be part of the `distr_dict`, whereas `y` and `z` are optional. So you could have
and your distr_dict is
```
{
'x' : scipy.stats.norm(0,1),
 'z' : scipy/.stats.uniform(1,2)
 }
```
Then `y=10`, and the other two follow their distribution.

Evelyn-M replied:
Ok, perfect.

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+        # zip_longest(flat_axes, cols, fillvalue=None):
+
+        nplots = len(self.distr_dict)
+        nrows, ncols = int(nplots / 3) + 1, min(nplots, 3)
```
Good idea, thanks!

Corrected in all the code in e69697a07dff258b54805e8acaf20f860c8a0f27

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+                                    fillvalue=None):
+            if name_distr is None:
+                ax.remove()
+                continue
```
When there is nothing to plot, the outer loop should continue without plotting (i.e. not go to line 132...), otherwise one would get an error. 
What this basically here does, is remove the extra canvases at the end. Say you have 4 distributions. Then you will have 2 rows of 3 plots. But the 5th and 6th panels would be empty plots. With the code here, I remove the axis entirely. I could also write the same as an if/else block. Would you prefer that?

Evelyn-M replied:
Ah, now I understand what the `continue` statement does. If line 129 evaluates to `True`, the `continue` makes the code continue with the next "item" in the for-loop instead of progressing to line 132.
Well.. It's neat in a sense, but I think a bit advanced. The if/else syntax is certainly easier to immediately understand for the average person looking at our source code, but we can also leave it this way, since it's not against any convention.

chahank replied:
Exactly! Is it too convoluted? I found it easier to read that big if...else.. blocks, but that depends on the familiarity with the `continue` statement. Should I change it?

Evelyn-M replied:
hmm. I don't know. I also agree that it's nice not to have huge semi-repetitive if/else blocks.. I think it's a tie with the fact that it is a bit convoluted if one doesnt know the continue syntax.

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+
+    def evaluate(self, uncvar_kwargs):
+        """
+        Evaluate the uncertainty variable.
```
It is true that it is not really needed. This is just a convenient method for `UncVar.unc_var(**uncvar_kwargs)`. It evaluates the function of the uncertainty variables at the given input parameters. I can remove it if you think it is clearer without?

Evelyn-M replied:
yes, i think it's clearer in the "classic" way you just described

chahank replied:
I will add that.

chahank replied:
Done here e33f4c96e6ebd590aeeb43fd69e2db841f93b73d

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+
+        Parameters
+        ----------
+        var : Object or else
```
Changed to `climada.uncertainty.UncVar or any other CLIMADA object`
bbfb8ede0221e2c7027cbd7c38b1fe0a28b8b455

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+            parameter names (all labels) from all unc_vars.
+            The default is pd.DataFrame().
+        metrics : dict(), optional
+            Dictionnary of the CLIMADA metrics for which the sensitivity
```
I  think this would be less precise, as not all attributes of `Impacts` or `CostBenefit` are fit for a sensitivity analysis, but only the outputs of the `.calc` methods. I called these outputs `metrics` (I heard this wording in a few talks on risk modelling and found it rather clear).

chahank replied:
Is it clearer now?

a0b71670c65e9384119256500fb64a267591cbad

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+            The default is pd.DataFrame().
+        metrics : dict(), optional
+            Dictionnary of the CLIMADA metrics for which the sensitivity
+            will be computed. For each sample, each metric must have a
```
I renamed `uncertainty.sample` to `uncertainty.samples_df` to make it clearer. Also, a sample is a row of `samples_df`. I changed the keyword argument to `samples`. Should I rather use `samples_df` too?

```
    def __init__(self, unc_vars, samples=None, metrics=None,
                 sensitivity=None):
```

57ad6342af1df6c5b2d91ef21922ad98b5d57cf1

Evelyn-M replied:
ah yes, more clear now! As for the kwarg, I personally like when the data type can be directly inferred from the name in the function definition. But up to you.

chahank replied:
Good idea! so `samples_df`, `metrics_dict` and `sensitivity_dict` or `si_dict` ?

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+            Keys are metric names (e.g. 'aai_agg', 'freq_curve') and 
+            values are pd.DataFrame with values for each parameter sample
+            (one row per sample).
+            The default is {}.
```
It is a bad idea to have default iterable values and against the CLIMADA conventions, c.f. https://github.com/CLIMADA-project/climada_python/blob/main/doc/guide/Guide_PythonDos-n-Donts.ipynb

I found it clearer to say in the docstring what the actual default value is rather than None.

chahank replied:
Hmmm... maybe this was not translated from the user guides prior to the coding convention update. It should be added then.

chahank replied:
Actually, just found the reference is here:

https://github.com/CLIMADA-project/climada_python/blob/guides_update/doc/guide/Guide_Reviewer_Checklist.ipynb

Evelyn-M replied:
Ok!

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+            The default is {}.
+        """
+    
+        self.unc_vars = unc_vars if unc_vars else {}
```
Not allowed, see above comment.

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+                             "correspond to parameters names of sample")
+        for metric, df_distr in self.metrics.items():
+            if df_distr.empty:
+                continue
```
Fixed in 05b2b181d0a6a6a5438395217b1819ec1d07d87c

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+            return self.sample.shape[0]
+        return 0
+
+    @property
+    def param_labels(self):
+        """
+        Labels of all uncertainty parameters.
+
+        Returns
+        -------
+        param_labels: list of strings
+            Labels of all uncertainty parameters.
+
+        """
+        return list(self.distr_dict.keys())
```
This is simply more flexible. If one can avoid unnecessary hard coding of variables that are actually a function of another variable, then one should not do it imho.

Evelyn-M replied:
Fair in general. This would mean the property is only assigned as an attribute to the parent class if the function which uses it, is actually called, right?
Two things that then speak against it: `param_labels` will always be used in the course of the uncertainty calculation, won't it? so it wouldn't unnecessarily "crowd" the `__init__` statement. Plus, assigning via decorators is extremely un-intuitive and makes it very very hard to directly see the structure / properties of a class for a first-time or not heavy user of this class. 

I think we've had the conversation about use of decorators earlier on in our coding convention meetings, with different opinions on that topic..

chahank replied:
>Fair in general. This would mean the property is only assigned as an attribute to the parent class if the function which uses it, is actually called, right?

That is not really correct. The attribute is always present, it just that the data is not saved in a separate memory slot, but retrieved from the already existing one (with maybe a computation in between).

One can of course define them in the `__init__`, but it is less flexible. Say you initiate your Object, then for some reason decide to add an uncertainty variable afterwards by modifying the `distr_dict`. For example, think of the case discussed with default values. You define the object, but then change `distr_dict = {'x':scipy.stats.norm(0,1)}` to `distri_dict = {'x':scipy.stats.norm(0,1), 'y':scipy.stats.norm(0,1)}`, then the code only works with the `@property`, otherwise note. I admit that this is not necessarily a common use case though.

Evelyn-M replied:
Right, i remember this from the pythonmorsel :D 
yeah well, alright then. I guess it has a certain advantage that may outweigh the readability downside, then.

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+        return distr_dict
+    
+    @property
+    def problem(self):
```
Good point. I was first trying to write the module such that it can also be compatible with something else than `salib`. I think the current version ended up not being that flexible anymore. I am a bit hesitant to introduce to many salib specific names, but maybe here it is indeed better.

chahank replied:
`salib_problem` or `problem_salib`?

Evelyn-M replied:
I have no preference on that. But actually, the link to the explanation should be different (see below)

chahank replied:
I changed it to `problem_sa` and update the link. Well spotted!
b2d7fce20822460ef5e8edaf317e8f0170f456db

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+            The default is 'saltelli'
+        sampling_kwargs: dict()
+            Optional keyword arguments of the chosen SALib sampling method.
+
```
Done here 

a078bab93331d3ce7149ced404b5895fe4c5c38f

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+
+    def calc_sensitivity(self, salib_method='sobol', method_kwargs=None):
+        """
+        Compute the sensitivity indices using SALib
```
Better? 

f8f75bb7bbd9879dfab732af487dfbafcb96d244

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+        for ax, col in zip_longest(flat_axes, cols, fillvalue=None):
+            if col is None:
+                ax.remove()
+                continue
```
The continue statement is so that the outer loop continues without executing the code after the if block.

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+   
+    def plot_sensitivity(self, salib_si='S1', metric_list=None):
+        """
+        Plot the first order sensitivity indices of the chosen metric.
```
One can plot any 1dimensional sensitivity index in principle, so in the case of Sobol it is "S1", "ST". The values from "S1_conf", "ST_conf" are the respective confidence intervals and are plotted as such. "S2" is a second order index which cannot be plotted as it is a matrix or correlations. I could add a matrix type plot for this case too.
Note that other sensitivity method will have other names for the sensitivity indices.

chahank replied:
Is the docstring clearer now?

8cc9eb537e5a527840ca867f3a30f593fa1c2fb4

chahank commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,728 @@
+        Parameters
+        ----------
+        salib_si: string, optional
+            The first order sensitivity index to plot (see SALib option)
+            https://salib.readthedocs.io/en/latest/basics.html
+            The default is S1.
```
The options depend on the chosen method, hence it is better to simply redirect to the actual documentation. One simply cannot use the sensitivity analysis without getting at least a bit familiar with salib and sensitvity analysis in general.

chahank replied:
See the updates 8cc9eb5 . Is it clearer now?

chahank commented on climada/engine/uncertainty/unc_cost_benefit.py:
```diff
@@ -0,0 +1,227 @@
+            Any keyword arguments of climada.engine.CostBenefit.calc()
+            EXCEPT: haz, ent, haz_fut, ent_fut
+
+        Returns
+        -------
+        None.
```
Good question. @emanuel-schmid : is it better that methods return `None` or nothing?

chahank commented on climada/engine/uncertainty/test/test_uncertainty.py:
```diff
@@ -0,0 +1,521 @@
+from pathos.pools import ProcessPool as Pool
+
+
+def impf_dem(x_impf=1):
```
Good point! Here the update:

1f0ae0a4d7db83194245ff25995249366137e268

chahank commented on climada/engine/uncertainty/unc_cost_benefit.py:
```diff
@@ -0,0 +1,227 @@
+
+        """
+
+        unc_vars = {
+            'haz': UncVar.var_to_uncvar(haz_unc),
+            'ent': UncVar.var_to_uncvar(ent_unc),
+            'haz_fut': UncVar.var_to_uncvar(haz_fut_unc),
+            'ent_fut': UncVar.var_to_uncvar(ent_fut_unc)
+            }
+
+        metrics =  {
+            'tot_climate_risk': pd.DataFrame([]),
+            'benefit': pd.DataFrame([]),
+            'cost_ben_ratio': pd.DataFrame([]),
+            'imp_meas_present': pd.DataFrame([]),
+            'imp_meas_future': pd.DataFrame([])
+            }
+
+        Uncertainty.__init__(self, unc_vars=unc_vars, metrics=metrics)
+
```
I think this is the cleaner way of initializing a new instance. But both should work.

chahank commented:
I think I addressed all the comments from @Evelyn-M . Thank you very much for the in-depth review! For several points I did not yet change anything in the code and will wait for feedback.

The tutorial will soon be raised to the standard of the rest of CLIMADA.

aleeciu commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,771 @@
+            (one row = one sample, columns = uncertainty parameters)
+        """
+        
+        self.samples_df = sampling_method
```
do you need the assignment at line 358? why is the df getting a string? Also I do not think is used before being reassigned at line 367

chahank replied:
Good catch, this line does not belong there.

aleeciu commented on climada/engine/uncertainty/unc_cost_benefit.py:
```diff
@@ -0,0 +1,227 @@
+         cost_ben_ratio] = list(zip(*cb_metrics))
+        elapsed_time = (time.time() - start) 
+        est_com_time = self.est_comp_time(elapsed_time, pool)
+        LOGGER.info(f"\n\nEstimated computation time: {est_com_time}s\n")
```
Why do you need the one_sample evaluation (block 118-127). Aren't you evaluating the full sample set just below - and also overwriting the 5 outputs? (Same in the Impact Uncertainty code)

chahank replied:
The idea is to get an estimate of the computation time by computing the output for one single sample, and then multiply by the number of samples.

aleeciu replied:
Mh ok.. why don't you just provide the time it takes to evaluate the full set? You could 1) drop those few lines, 2) provide the exact computation as it may not scale linearly, 3) you need to compute it ayway...

chahank replied:
Well, that is then a very different use case. The goal here is to warn the user if the computation is longer than usual. So far, every person that used this model made very simple basic mistakes that changed the computation time from a few seconds/minutes to several hours/days. Hence, my idea to always give a quick estimate of the total computation time (it is a very rough one, but enough to catch these very very bad mistakes), and suggest one of the major problems I spotted so far. 

If someone wants to do a scaling analysis, they should do it themselves.

aleeciu replied:
Ok now it makes more sense..

chahank replied:
If you have a more elegant solution to that, please let me know!

aleeciu replied:
if your goal is to warn the user about the total computation time in order for her to quit the process in case it takes too long.. then I think it makes sense to do it the way you do it... I just did not see the point but now I do

aleeciu commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,771 @@
+    --------
+
+    Categorical variable function: LitPop exposures with m,n exponents in [0,5]
+        import scipy as sp
```
Just wondering whether it would make sense to have some subclasses which support building the type of funcs at lines 47 to 77 - one for Haz, one for LitPop, one for IF; so that the user does not have to do it himself if he does not know how. But this is an advanced functionality so we also assume it will not be used by the average user.

chahank replied:
I agree that this could be an idea. I would advocate to first gather some experience with users to see what is needed.

aleeciu commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,771 @@
+
+        """
+        return self.uncvar_func(**uncvar_kwargs)
+    
```
trailing space - also in many other places...

chahank replied:
Thanks, will remove them!

aleeciu commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,771 @@
+        
+        #Certaint Salib method required model input (X) and output (Y), others
+        #need only ouput (Y)
+        salib_kwargs = method.analyze.__code__.co_varnames
```
I think I understand what you are doing, i.e., getting the full set of possible kwargs for the selected salib method (right?); but as this is not common (at least it is not for me) maybe it is good to say it explicitely in the comment just above.

chahank replied:
I added a comment in 6c66e6729cf297e478eb6501d835ba3bdc26c0e5

aleeciu commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,771 @@
+        
+        self.sensitivity = sensitivity_dict
+
+        return sensitivity_dict
```
From the tutorial it seems the output dict of such function is of the form (for e.g., impact):
`
{'aai_agg': {'aai_agg': sensitivity_indexes_dict},
 'freq_curve': {'rp50': sensitivity_indexes_dict, 'rp100': sensitivity_indexes_dict, 
'rp250': sensitivity_indexes_dict},
 'tot_value': {'tot_value': sensitivity_indexes_dict}}
`
why not something like:
`
{'aai_agg': sensitivity_indexes_dict,
 'rp50': sensitivity_indexes_dict, 'rp100': sensitivity_indexes_dict, 
'rp250': sensitivity_indexes_dict, 'tot_value': sensitivity_indexes_dict}
`
I think it is much easier to ready and, to me, there is no need to group the RPs under the key freq_curve, especially if this requires redundant info for all other keys.

chahank replied:
Good point. The reason I did it this way is for simplicity and compatibility with the cost benefit part. Basically, the keys now are directly the keys that come out of the `impact.calc` and `costbenefit.calc`. These are dictionnaries. In the case of impact, it looks a bit awkward, but for the cost benefit it makes sense to me. 

For example, `cost_benefit.calc()` produces the dictionnaries 
```
        imp_meas_future (dict): impact of each measure at future or default.
            Key: measure name ('no measure' used for case without measure),
            Value: dict with:
                             'cost' (tuple): (cost measure, cost factor insurance),
                             'risk' (float): risk measurement,
                             'risk_transf' (float): annual expected risk transfer,
                             'efc'  (ImpactFreqCurve): impact exceedance freq
                (optional)   'impact' (Impact): impact instance
        imp_meas_present (dict): impact of each measure at present.
            Key: measure name ('no measure' used for case without measure),
            Value: dict with:
                             'cost' (tuple): (cost measure, cost factor insurance),
                             'risk' (float): risk measurement,
                             'risk_transf' (float): annual expected risk transfer,
                             'efc'  (ImpactFreqCurve): impact exceedance freq
                (optional)   'impact' (Impact): impact instance
```

Both of these have the same keys, twice! Hence, I found it simpler to have nested dictionaries that directly take the outputs of the `.calc`.

aleeciu reviewed (commented):
Well done - I had few minor points; after that this is ready to go in develop and main as far as I am concern. Also well done with keeping the saltelli sampling for the base sample [0,1] and only after applying any desired pdf.. for some reason Salib developers did not do it and thus only allows a bunch of two-params distributions to be used (i.e., 'unif', 'triang', 'norm', 'lognorm' - https://github.com/SALib/SALib/blob/master/src/SALib/util/__init__.py.. hope we aren't missing something though).

chahank commented:
>  Also well done with keeping the saltelli sampling for the base sample [0,1] and only after applying any desired pdf.. for some reason Salib developers did not do it and thus only allows a bunch of two-params distributions to be used (i.e., 'unif', 'triang', 'norm', 'lognorm' - https://github.com/SALib/SALib/blob/master/src/SALib/util/__init__.py.. hope we aren't missing something though).

This is actually a point we might have to discuss in more details. What I understood from the SAlib forum is that adjusting for the distribution afterwards as is done now in the uncertainty module reduces the accuracy of the sensitivity analysis (i.e., more samples are needed for convergence). In the worst case, the data would not converge. I am however not sure how important this issue is. In any case SAlib returns the standard deviation values which means one at least has feedback whether the algorithm converged or not. I thought that for the time being it is fine in the context of CLIMADA because 1) most cases will use uniform distributions 2) it is simple and practicable 3) having a high number of samples in CLIMADA is not that problematic.

chahank commented:
The tutorial has now been rewritten entirely and should explain better how to use the module. Maybe I will add one more section to explain the Uncertainty base class and how to manually construct a sample or distribution for the uncertainty / sensitivity analysis.

aleeciu commented on climada/engine/uncertainty/base.py:
```diff
@@ -197,39 +199,39 @@ def __init__(self, unc_vars, samples=None, metrics=None,
             The default is pd.DataFrame().
         metrics : dict(), optional
-            Dictionnary of the CLIMADA metrics (outputs from 
+            Dictionnary of the CLIMADA metrics (outputs from
```
minor point: "dictionary" is spelled with one n... also in few other places... please check

chahank replied:
O no, again the same mistake. Thanks!

chahank commented:
Thanks to @aleeciu and @Evelyn-M  for the reviews! 

While writing the tutorial, I realized that the name 'distribution' for the values of the metrics for each sample might be confusing, as there is also the distribution of the uncertainty parameters. Hence, do you think I should change the name `.calc_distribution` to for example `.calc_metrics_distribution` or `.calc_metrics` or else?

emanuel-schmid commented on requirements/env_climada.yml:
```diff
@@ -25,6 +25,7 @@ dependencies:
   - python=3.8
   - rasterio>=1.1
   - requests>=2.24
+  - salib>=1.3.0
```
```suggestion
```

emanuel-schmid commented on requirements/env_climada.yml:
```diff
@@ -25,6 +25,7 @@ dependencies:
   - python=3.8
   - rasterio>=1.1
   - requests>=2.24
+  - salib>=1.3.0
```
I'd move this to the pip section. In any case it will be installed through pip.

chahank replied:
It can be installed via conda at the moment at least.

emanuel-schmid commented:
> For instance, the module is already 984 lines long (with all coments etc). But it is expected that more features will be added. Is this ok or not? @emanuel-schmid is it ok to add the dependency `SALib` to CLIMADA?

No objections, I've already added it in the develop branch. (In the pip section though.)

chahank commented:
Quick update: I added the following
1)

Evelyn-M commented on climada/engine/uncertainty/base.py:
```diff
@@ -0,0 +1,849 @@
+        """
+        The description of the uncertainty variables and their
+        distribution as used in SALib.
+        https://salib.readthedocs.io/en/latest/getting-started.html.
```
```suggestion
        https://salib.readthedocs.io/en/latest/basics.html
```

chahank replied:
Already done in b2d7fce20822460ef5e8edaf317e8f0170f456db :)

Evelyn-M commented on climada/util/value_representation.py:
```diff
@@ -29,17 +29,21 @@
-    Examples: 1.234567 -> 1.2346, 123456.89 -> 123460.0
+    0, inf, Nan are returned unchanged.
+    Examples: n_sig_dig = 5
+        1.234567 -> 1.2346, 123456.89 -> 123460.0
```
I cannot find my previous comment now to this, so maybe I cross-post here:
why does the sig_dig in this example correctly round up the ...56.89 to ...60.0, but wrongly rounds down 1.234567 to 1.2346?

chahank replied:
This is the expected behaviour. This returns a certain number of "significant digits", it does not round. 

For ` 1.234567`, the first 5 digits are ` 1.2345`. Then, the last one is rounded up because the 6th digit is `6>5`. Al the remaining digits are set to 0.

Evelyn-M replied:
ah ah, my bad. i mis-read it.

Evelyn-M replied:
> What do you mean?
I don't understand what you don't understand, I'm afraid.

chahank replied:
```
impact_metrics = self.samples_df.apply(_map_impact_calc) ?
```
I still think the mapping cannot be done in this syntax, be I will give it a shot since it would more elegant.

chahank replied:
OK, I think this does not work. For the implementation, I need to access the `self.calc_eai_exp` and `self.calc_at_event`. However, if I use `self.samples_df.apply(_map_impact_calc)`, the `self` si not passed. I tried it a bit, but did not succeed. I do not really see a good reason to change that either. So, unless you insists (which is your right), I would leave it as is.

chahank commented:
In the latest commit 1e231f2526d7f741004b060e4cb7f9bdcb7bdb72, I added save/load functionalities to the module. Please have a look whether this is useful or not, otherwise I can revert this commit easily.

Evelyn-M commented:
Alright. I don't see any major things anymore that would impede a merge ;)! you have my blessing^^

chahank commented:
Thanks a lot for the in-depth reviews @Evelyn-M  and @aleeciu and @emanuel-schmid  !!
